// ngx-responsive-image/esm5/lib/media/media.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BreakpointObserver } from '@angular/cdk/layout';
import { Inject, Injectable } from '@angular/core';
import { merge } from 'rxjs';
import { distinctUntilChanged, filter, first, map, pairwise, shareReplay, startWith, takeWhile } from 'rxjs/operators';
import { BREAKPOINTS } from '../breakpoints.token';
import { IMAGE_WIDTHS } from '../image-widths.token';
var MediaService = /** @class */ (function () {
    function MediaService(breakpointObserver, breakpoints, imageWidths) {
        var _this = this;
        this.breakpointObserver = breakpointObserver;
        this.breakpoints = breakpoints;
        this.imageWidths = imageWidths;
        this.breakpoints$ = this.breakpointObserver.observe(this.breakpoints);
        this.breakpointChange$ = this.breakpoints$.pipe(filter((/**
         * @param {?} breakpointState
         * @return {?}
         */
        function (breakpointState) { return breakpointState.matches; })), distinctUntilChanged((/**
         * @param {?} previous
         * @param {?} current
         * @return {?}
         */
        function (previous, current) {
            return Object.keys(previous.breakpoints)[Object.values(previous.breakpoints).indexOf(true)] ===
                Object.keys(current.breakpoints)[Object.values(current.breakpoints).indexOf(true)];
        })), startWith((/** @type {?} */ (null))), pairwise());
        this.breakpointUp$ = this.breakpointChange$.pipe(filter((/**
         * @param {?} __0
         * @param {?} i
         * @return {?}
         */
        function (_a, i) {
            var _b = tslib_1.__read(_a, 2), previous = _b[0], current = _b[1];
            if (i === 0) {
                return true;
            }
            else {
                /** @type {?} */
                var previousBreakpoints = Object.values(previous.breakpoints);
                /** @type {?} */
                var currentBreakpoints = Object.values(current.breakpoints);
                /** @type {?} */
                var pbi = previousBreakpoints.findIndex((/**
                 * @param {?} breakpointValue
                 * @return {?}
                 */
                function (breakpointValue) { return breakpointValue; }));
                /** @type {?} */
                var cbi = currentBreakpoints.findIndex((/**
                 * @param {?} breakpointValue
                 * @return {?}
                 */
                function (breakpointValue) { return breakpointValue; }));
                return cbi > pbi;
            }
        })), filter((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), previous = _b[0], current = _b[1];
            /** @type {?} */
            var currentBkreakpoint = Object.keys(current.breakpoints)[Object.values(current.breakpoints).indexOf(true)];
            if (!_this.maximumBreakpoint ||
                Object.keys(current.breakpoints).indexOf(currentBkreakpoint) >
                    Object.keys(current.breakpoints).indexOf(_this.maximumBreakpoint)) {
                _this.maximumBreakpoint = currentBkreakpoint;
                return true;
            }
            else {
                return false;
            }
        })), map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), previous = _b[0], current = _b[1];
            return Object.keys(current.breakpoints)[Object.values(current.breakpoints).indexOf(true)];
        })), takeWhile((/**
         * @param {?} currentBreakpoint
         * @return {?}
         */
        function (currentBreakpoint) {
            return (_this.maximumBreakpoint !== _this.breakpoints[_this.breakpoints.length - 1]);
        }), true), shareReplay());
        this.breakpointAndWidthUp$ = merge(this.breakpoints$.pipe(first(), map((/**
         * @param {?} breakpointState
         * @return {?}
         */
        function (breakpointState) {
            return Object.keys(breakpointState.breakpoints)[Object.values(breakpointState.breakpoints).indexOf(true)];
        }))), this.breakpointUp$).pipe(map((/**
         * @param {?} breakpoint
         * @return {?}
         */
        function (breakpoint) {
            return {
                width: _this.imageWidths[_this.breakpoints.indexOf(breakpoint)],
                breakpoint: breakpoint
            };
        })), shareReplay());
    }
    MediaService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MediaService.ctorParameters = function () { return [
        { type: BreakpointObserver },
        { type: undefined, decorators: [{ type: Inject, args: [BREAKPOINTS,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [IMAGE_WIDTHS,] }] }
    ]; };
    return MediaService;
}());
export { MediaService };
if (false) {
    /** @type {?} */
    MediaService.prototype.breakpoints$;
    /** @type {?} */
    MediaService.prototype.breakpointChange$;
    /** @type {?} */
    MediaService.prototype.breakpointUp$;
    /** @type {?} */
    MediaService.prototype.breakpointAndWidthUp$;
    /**
     * @type {?}
     * @private
     */
    MediaService.prototype.maximumBreakpoint;
    /**
     * @type {?}
     * @private
     */
    MediaService.prototype.breakpointObserver;
    /**
     * @type {?}
     * @private
     */
    MediaService.prototype.breakpoints;
    /**
     * @type {?}
     * @private
     */
    MediaService.prototype.imageWidths;
}
