// ngx-responsive-image/esm2015/lib/responsive-image/responsive-image.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isPlatformBrowser } from '@angular/common';
import { Directive, ElementRef, EventEmitter, Inject, Input, Output, PLATFORM_ID, Renderer2 } from '@angular/core';
import { MediaService } from '../media/media.service';
export class ResponsiveImageDirective {
    /**
     * @param {?} platformId
     * @param {?} element
     * @param {?} mediaService
     * @param {?} renderer2
     */
    constructor(platformId, element, mediaService, renderer2) {
        this.platformId = platformId;
        this.element = element;
        this.mediaService = mediaService;
        this.renderer2 = renderer2;
        this.breakpointUp = new EventEmitter();
        this.subscriptions = {};
    }
    /**
     * @param {?} simpleChange
     * @return {?}
     */
    ngOnChanges(simpleChange) {
        if (isPlatformBrowser(this.platformId)) {
            if (this.subscriptions.mediaSubscription) {
                this.subscriptions.mediaSubscription.unsubscribe();
            }
            if (this.imgSrc) {
                this.subscriptions.mediaSubscription = this.mediaService.breakpointAndWidthUp$.subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                result => {
                    if (this.manual) {
                        this.breakpointUp.emit({
                            imgSrc: this.imgSrc,
                            breakpoint: result.breakpoint,
                            width: result.width
                        });
                    }
                    else {
                        this.renderer2.setAttribute(this.element.nativeElement, 'src', this.imgSrc.replace(':width', result.width));
                    }
                }));
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        Object.keys(this.subscriptions).forEach((/**
         * @param {?} sk
         * @return {?}
         */
        sk => this.subscriptions[sk].unsubscribe()));
    }
}
ResponsiveImageDirective.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line: directive-selector
                selector: 'img[responsiveImage]'
            },] }
];
/** @nocollapse */
ResponsiveImageDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
    { type: ElementRef },
    { type: MediaService },
    { type: Renderer2 }
];
ResponsiveImageDirective.propDecorators = {
    imgSrc: [{ type: Input }],
    manual: [{ type: Input }],
    breakpointUp: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ResponsiveImageDirective.prototype.imgSrc;
    /** @type {?} */
    ResponsiveImageDirective.prototype.manual;
    /** @type {?} */
    ResponsiveImageDirective.prototype.breakpointUp;
    /**
     * @type {?}
     * @private
     */
    ResponsiveImageDirective.prototype.subscriptions;
    /**
     * @type {?}
     * @private
     */
    ResponsiveImageDirective.prototype.platformId;
    /**
     * @type {?}
     * @private
     */
    ResponsiveImageDirective.prototype.element;
    /**
     * @type {?}
     * @private
     */
    ResponsiveImageDirective.prototype.mediaService;
    /**
     * @type {?}
     * @private
     */
    ResponsiveImageDirective.prototype.renderer2;
}
