// ngx-responsive-image/fesm5/ngx-responsive-image.js
import { InjectionToken, Injectable, Inject, Directive, PLATFORM_ID, ElementRef, Renderer2, Input, Output, EventEmitter, NgModule } from '@angular/core';
import { Breakpoints, BreakpointObserver } from '@angular/cdk/layout';
import { __read } from 'tslib';
import { merge } from 'rxjs';
import { filter, distinctUntilChanged, startWith, pairwise, map, takeWhile, shareReplay, first } from 'rxjs/operators';
import { ObserversModule } from '@angular/cdk/observers';
import { isPlatformBrowser, CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var BREAKPOINTS = new InjectionToken('BREAKPOINTS');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var DEFAULT_BREAKPOINTS = [
    Breakpoints.XSmall,
    Breakpoints.Small,
    Breakpoints.Medium,
    Breakpoints.Large,
    Breakpoints.XLarge
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var DEFAULT_WIDTHS = [300, 600, 960, 1280, 1920];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var IMAGE_WIDTHS = new InjectionToken('IMAGE_WIDTHS');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var MediaService = /** @class */ (function () {
    function MediaService(breakpointObserver, breakpoints, imageWidths) {
        var _this = this;
        this.breakpointObserver = breakpointObserver;
        this.breakpoints = breakpoints;
        this.imageWidths = imageWidths;
        this.breakpoints$ = this.breakpointObserver.observe(this.breakpoints);
        this.breakpointChange$ = this.breakpoints$.pipe(filter((/**
         * @param {?} breakpointState
         * @return {?}
         */
        function (breakpointState) { return breakpointState.matches; })), distinctUntilChanged((/**
         * @param {?} previous
         * @param {?} current
         * @return {?}
         */
        function (previous, current) {
            return Object.keys(previous.breakpoints)[Object.values(previous.breakpoints).indexOf(true)] ===
                Object.keys(current.breakpoints)[Object.values(current.breakpoints).indexOf(true)];
        })), startWith((/** @type {?} */ (null))), pairwise());
        this.breakpointUp$ = this.breakpointChange$.pipe(filter((/**
         * @param {?} __0
         * @param {?} i
         * @return {?}
         */
        function (_a, i) {
            var _b = __read(_a, 2), previous = _b[0], current = _b[1];
            if (i === 0) {
                return true;
            }
            else {
                /** @type {?} */
                var previousBreakpoints = Object.values(previous.breakpoints);
                /** @type {?} */
                var currentBreakpoints = Object.values(current.breakpoints);
                /** @type {?} */
                var pbi = previousBreakpoints.findIndex((/**
                 * @param {?} breakpointValue
                 * @return {?}
                 */
                function (breakpointValue) { return breakpointValue; }));
                /** @type {?} */
                var cbi = currentBreakpoints.findIndex((/**
                 * @param {?} breakpointValue
                 * @return {?}
                 */
                function (breakpointValue) { return breakpointValue; }));
                return cbi > pbi;
            }
        })), filter((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = __read(_a, 2), previous = _b[0], current = _b[1];
            /** @type {?} */
            var currentBkreakpoint = Object.keys(current.breakpoints)[Object.values(current.breakpoints).indexOf(true)];
            if (!_this.maximumBreakpoint ||
                Object.keys(current.breakpoints).indexOf(currentBkreakpoint) >
                    Object.keys(current.breakpoints).indexOf(_this.maximumBreakpoint)) {
                _this.maximumBreakpoint = currentBkreakpoint;
                return true;
            }
            else {
                return false;
            }
        })), map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = __read(_a, 2), previous = _b[0], current = _b[1];
            return Object.keys(current.breakpoints)[Object.values(current.breakpoints).indexOf(true)];
        })), takeWhile((/**
         * @param {?} currentBreakpoint
         * @return {?}
         */
        function (currentBreakpoint) {
            return (_this.maximumBreakpoint !== _this.breakpoints[_this.breakpoints.length - 1]);
        }), true), shareReplay());
        this.breakpointAndWidthUp$ = merge(this.breakpoints$.pipe(first(), map((/**
         * @param {?} breakpointState
         * @return {?}
         */
        function (breakpointState) {
            return Object.keys(breakpointState.breakpoints)[Object.values(breakpointState.breakpoints).indexOf(true)];
        }))), this.breakpointUp$).pipe(map((/**
         * @param {?} breakpoint
         * @return {?}
         */
        function (breakpoint) {
            return {
                width: _this.imageWidths[_this.breakpoints.indexOf(breakpoint)],
                breakpoint: breakpoint
            };
        })), shareReplay());
    }
    MediaService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MediaService.ctorParameters = function () { return [
        { type: BreakpointObserver },
        { type: undefined, decorators: [{ type: Inject, args: [BREAKPOINTS,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [IMAGE_WIDTHS,] }] }
    ]; };
    return MediaService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ResponsiveImageDirective = /** @class */ (function () {
    function ResponsiveImageDirective(platformId, element, mediaService, renderer2) {
        this.platformId = platformId;
        this.element = element;
        this.mediaService = mediaService;
        this.renderer2 = renderer2;
        this.breakpointUp = new EventEmitter();
        this.subscriptions = {};
    }
    /**
     * @param {?} simpleChange
     * @return {?}
     */
    ResponsiveImageDirective.prototype.ngOnChanges = /**
     * @param {?} simpleChange
     * @return {?}
     */
    function (simpleChange) {
        var _this = this;
        if (isPlatformBrowser(this.platformId)) {
            if (this.subscriptions.mediaSubscription) {
                this.subscriptions.mediaSubscription.unsubscribe();
            }
            if (this.imgSrc) {
                this.subscriptions.mediaSubscription = this.mediaService.breakpointAndWidthUp$.subscribe((/**
                 * @param {?} result
                 * @return {?}
                 */
                function (result) {
                    if (_this.manual) {
                        _this.breakpointUp.emit({
                            imgSrc: _this.imgSrc,
                            breakpoint: result.breakpoint,
                            width: result.width
                        });
                    }
                    else {
                        _this.renderer2.setAttribute(_this.element.nativeElement, 'src', _this.imgSrc.replace(':width', result.width));
                    }
                }));
            }
        }
    };
    /**
     * @return {?}
     */
    ResponsiveImageDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var _this = this;
        Object.keys(this.subscriptions).forEach((/**
         * @param {?} sk
         * @return {?}
         */
        function (sk) {
            return _this.subscriptions[sk].unsubscribe();
        }));
    };
    ResponsiveImageDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line: directive-selector
                    selector: 'img[responsiveImage]'
                },] }
    ];
    /** @nocollapse */
    ResponsiveImageDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: ElementRef },
        { type: MediaService },
        { type: Renderer2 }
    ]; };
    ResponsiveImageDirective.propDecorators = {
        imgSrc: [{ type: Input }],
        manual: [{ type: Input }],
        breakpointUp: [{ type: Output }]
    };
    return ResponsiveImageDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var NgxResponsiveImageModule = /** @class */ (function () {
    function NgxResponsiveImageModule() {
    }
    /**
     * @param {?} breakpoints
     * @param {?} imageWidths
     * @return {?}
     */
    NgxResponsiveImageModule.forRoot = /**
     * @param {?} breakpoints
     * @param {?} imageWidths
     * @return {?}
     */
    function (breakpoints, imageWidths) {
        return {
            ngModule: NgxResponsiveImageModule,
            providers: [
                {
                    provide: BREAKPOINTS,
                    useValue: breakpoints
                },
                {
                    provide: IMAGE_WIDTHS,
                    useValue: imageWidths
                },
                MediaService
            ]
        };
    };
    NgxResponsiveImageModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ResponsiveImageDirective],
                    imports: [ObserversModule, CommonModule],
                    exports: [ResponsiveImageDirective]
                },] }
    ];
    return NgxResponsiveImageModule;
}());

export { BREAKPOINTS, DEFAULT_BREAKPOINTS, DEFAULT_WIDTHS, IMAGE_WIDTHS, MediaService, NgxResponsiveImageModule, ResponsiveImageDirective };
