// ngx-responsive-image/esm5/lib/ngx-responsive-image.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ObserversModule } from '@angular/cdk/observers';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { BREAKPOINTS } from './breakpoints.token';
import { IMAGE_WIDTHS } from './image-widths.token';
import { MediaService } from './media/media.service';
import { ResponsiveImageDirective } from './responsive-image/responsive-image.directive';
var NgxResponsiveImageModule = /** @class */ (function () {
    function NgxResponsiveImageModule() {
    }
    /**
     * @param {?} breakpoints
     * @param {?} imageWidths
     * @return {?}
     */
    NgxResponsiveImageModule.forRoot = /**
     * @param {?} breakpoints
     * @param {?} imageWidths
     * @return {?}
     */
    function (breakpoints, imageWidths) {
        return {
            ngModule: NgxResponsiveImageModule,
            providers: [
                {
                    provide: BREAKPOINTS,
                    useValue: breakpoints
                },
                {
                    provide: IMAGE_WIDTHS,
                    useValue: imageWidths
                },
                MediaService
            ]
        };
    };
    NgxResponsiveImageModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ResponsiveImageDirective],
                    imports: [ObserversModule, CommonModule],
                    exports: [ResponsiveImageDirective]
                },] }
    ];
    return NgxResponsiveImageModule;
}());
export { NgxResponsiveImageModule };
